src/Analysis/Ast/Impl/Analyzer/PythonAnalyzer.cs

```diff
@@ -281,19 +296,97 @@ private void StartAnalysis(IDependencyChainNode<PythonAnalyzerEntry> node, int v
                 var analysis = new DocumentAnalysis((IDocument)module, version, walker.GlobalScope, walker.Eval);
 
                 (module as IAnalyzable)?.NotifyAnalysisComplete(analysis);
-                node.Value.TrySetAnalysis(analysis, version, _syncObj);
+                entry.TrySetAnalysis(analysis, version);
+                node.Commit();
 
-                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) complete in {(stopWatch.Elapsed - startTime).TotalMilliseconds} ms.");
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) completed in {(stopWatch.Elapsed - startTime).TotalMilliseconds} ms.");
             } catch (OperationCanceledException oce) {
-                node.Value.TryCancel(oce, version, _syncObj);
+                node.Value.TryCancel(oce, version);
+                node.Skip();
+                var module = node.Value.Module;
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) canceled.");
             } catch (Exception exception) {
-                node.Value.TrySetException(exception, version, _syncObj);
+                var module = node.Value.Module;
+                node.Value.TrySetException(exception, version);
+                node.Commit();
+                _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) failed.");
             } finally {
                 Interlocked.Decrement(ref _runningTasks);
-                node.Commit();
             }
         }
 
+        private bool TryFindDependencies(PythonAnalyzerEntry entry, int version, CancellationToken cancellationToken, out ImmutableArray<ModuleKey> dependencies) {
+            IPythonModule module;
+            PythonAst ast;
+            lock (_syncObj) {
+                if (!entry.IsValidVersion(version, out module, out ast)) {
+                    dependencies = ImmutableArray<ModuleKey>.Empty;
+                    return false;
+                }
+            }
+
+            var dependenciesHashSet = new HashSet<ModuleKey>();
+            var isTypeshed = module is StubPythonModule stub && stub.IsTypeshed;
+            var moduleResolution = module.Interpreter.ModuleResolution;
+            var pathResolver = isTypeshed
+                ? module.Interpreter.TypeshedResolution.CurrentPathResolver
+                : moduleResolution.CurrentPathResolver;
+
+            if (module.Stub != null) {
+                dependenciesHashSet.Add(new ModuleKey(module.Stub));
```

Comment:
Do we need stubs as dependencies? I think we used to forcibly analyze and attach them at import time without going through actual dependency analysis.

Reply:
Stub must be analyzed before the module. So graph must be aware of this dependency. Preserving this dependency is important in cases when stub has its own analysis dependency on another stub.